source/common/grpc/buffered_async_client.h

```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+#include <cstdint>
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/container/btree_map.h"
+
+namespace Envoy {
+namespace Grpc {
+
+enum class BufferState { Buffered, PendingFlush };
+
+// This class wraps bidirectional gRPC and provides message arrival guarantee.
+// It stores messages to be sent or in the process of being sent in a buffer,
+// and can track the status of the message based on the ID assigned to each message.
+// If a message fails to be sent, it can be re-buffered to guarantee its arrival.
+template <class RequestType, class ResponseType> class BufferedAsyncClient {
+public:
+  BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
+                      Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+      : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
+        client_(client) {}
+
+  ~BufferedAsyncClient() {
+    if (active_stream_ != nullptr) {
+      active_stream_ = nullptr;
+    }
+  }
+
+  absl::optional<uint64_t> bufferMessage(RequestType& message) {
+    const auto buffer_size = message.ByteSizeLong();
+    if (current_buffer_bytes_ + buffer_size > max_buffer_bytes_) {
+      return absl::nullopt;
+    }
+
+    auto id = publishId();
+    message_buffer_[id] = std::make_pair(BufferState::Buffered, message);
+    current_buffer_bytes_ += buffer_size;
+    return id;
+  }
+
+  absl::flat_hash_set<uint64_t> sendBufferedMessages() {
+    if (active_stream_ == nullptr) {
+      active_stream_ =
+          client_.start(service_method_, callbacks_, Http::AsyncClient::StreamOptions());
+    }
+
+    if (active_stream_->isAboveWriteBufferHighWatermark()) {
+      return {};
+    }
+
+    absl::flat_hash_set<uint64_t> inflight_message_ids;
+
+    for (auto&& it : message_buffer_) {
+      const auto id = it.first;
+      auto& state = it.second.first;
+      auto& message = it.second.second;
+
+      if (state == BufferState::PendingFlush) {
+        continue;
```

Comment:
I think we should add a test that covers this case, since it prevents double sending. I noticed it's not hit in the [coverage report](https://storage.googleapis.com/envoy-pr/3ec75b0/coverage/source/common/grpc/buffered_async_client.h.gcov.html#63)

Reply:
added